### PR TITLE
sophosxg-module initial release

### PR DESCRIPTION
**Pull request by StefanSa**
## What does this PR do?
This PR Introduces the SophosXG filebeat module. Focusing currently on SophosXG Firewall, but should include other Sophos products as separate PR's later on.

## Why is it important?
Adding more supported products to the filebeat portfolio.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally
`INTEGRATION_TESTS=1 BEAT_STRICT_PERMS=false TESTING_FILEBEAT_MODULES=sophosxg nosetests -v -s tests/system/test_xpack_modules.py`

### CODE REVIEW

**Comment by P1llus** [comment]
This is a PR that was planned with @StefanSa over email, we will try to collaborate and apply any changes needed before we can start a review.

Thanks a lot for the contribution!

**Comment by P1llus** [comment]
cla/check

**Comment by adriansr** [comment]
jenkins, test this

**Comment by adriansr** [comment]
Jenkins, run tests please

**Comment by P1llus** [comment]
Jenkins, run tests please

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        Starttime
+
+    - name: remote_ip
+      type: keyword
+      description: >
+        Remote IP
+
+    - name: timestamp
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:744]: type: date

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        Server
+
+    - name: host
+      type: keyword
+      description: >
+        Host
+
+    - name: responsetime
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:774]: type: date?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        extra
+
+    - name: PHPSESSID
+      type: keyword
+      description: >
+        PHPSESSID
+
+    - name: start_time
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:799]: type: date?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        Start time
+
+    - name: eventtime
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:804]: type: date?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 5]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        RED ID
+
+    - name: branch_name
+      type: keyword
+      description: >
+        Branch Name
+
+    - name: updatedip
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:819]: type: ip?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 6]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        transmitted errors
+
+    - name: receivederrors
+      type: keyword
+      description: >
+        received errors
+
+    - name: receivedkbits
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:869]: type: long?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 7]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        received kbits
+
+    - name: transmittedkbits
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:874]: type: long?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 8]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        transmitted kbits
+
+    - name: transmitteddrops
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:879]: type: long?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 9]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        transmitted drops
+
+    - name: receiveddrops
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:884]: type: long?

File: x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml [hunk 10]
```diff
@@ -0,0 +1,937 @@
...
+      description: >
+        received drops
+
+    - name: collisions
+      type: keyword
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/_meta/fields.yml:889]: type: long?

File: x-pack/filebeat/module/sophosxg/firewall/ingest/event.yml [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+processors:
+#######################
+## ECS Event Mapping ##
+#######################
+- set:
+    field: event.kind
+    value: event
+- set:
+    field: event.outcome
+    value: succes
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/event.yml:11]: fix spelling for "success"

File: x-pack/filebeat/module/sophosxg/firewall/ingest/event.yml [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    if: 'ctx?.sophosxg?.firewall?.log_subtype == "Authentication" && ctx?.sophosxg?.firewall?.status == "Successful"'
+- set:
+    field: event.outcome
+    value: failure
+    if: 'ctx?.sophosxg?.firewall?.log_subtype == "Authentication" && ctx?.sophosxg?.firewall?.status == "Failed"'
+- set:
+    field: event.outcome
+    value: succes
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/event.yml:19]: fix spelling for "success"

File: x-pack/filebeat/module/sophosxg/firewall/ingest/firewall.yml [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+    field: event.type
+    value: 
+      - end
+      - allowed
+      - connection
+    if: "ctx.sophosxg?.firewall?.connevent == 'Stop'"
+- append:
+    field: event.type
+    value:
+      - diened
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/firewall.yml:45]: fix spelling "denied"

File: x-pack/filebeat/module/sophosxg/firewall/ingest/firewall.yml [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+    value: '{{destination.ip}}'
+- append:
+    if: 'ctx?.source?.nat?.ip != null'
+    field: related.ip
+    value: '{{source.nat.ip}}'
+- append:
+    if: 'ctx?.destination?.nat?.ip != null'
+    field: related.ip
+    value: '{{destination.nat.ip}}'
+
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/firewall.yml:413]: might want to append source.user.name to related.user as well.

File: x-pack/filebeat/module/sophosxg/firewall/ingest/sandstorm.yml [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+- append:
+    field: event.type
+    value: 
+      - end
+      - connection
+    if: "ctx.sophosxg?.firewall?.reason == 'eligible'"
+- append:
+    field: event.type
+    value:
+      - diened
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/sandstorm.yml:50]: fix spelling "denied"

File: x-pack/filebeat/module/sophosxg/firewall/ingest/waf.yml [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+- append:
+    field: event.type
+    value:
+      - allowed
+      - connection
+    if: 'ctx.sophosxg?.firewall?.reason == "-"'
+- append:
+    field: event.type
+    value:
+      - diened
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/waf.yml:46]: fix spelling "denied"

File: x-pack/filebeat/module/sophosxg/firewall/ingest/waf.yml [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+    field: related.user
+    value: "{{source.user.name}}"
+    if: "ctx.source?.user?.name != null"
+
+#############
+## Cleanup ##
+#############
+- lowercase:
+      field: http.request.method
+      ignore_failure: true
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/waf.yml:266]: not required by ECS anymore, new guidance is to leave case alone.  I'd suggest removing this block.

File: x-pack/filebeat/module/sophosxg/firewall/ingest/wifi.yml [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+processors:
+#######################
+## ECS Event Mapping ##
+#######################
+- set:
+    field: event.kind
+    value: event
+- set:
+    field: event.outcome
+    value: succes
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/ingest/wifi.yml:11]: fix spelling "success"

File: x-pack/filebeat/module/sophosxg/firewall/test/sandbox.log-expected.json [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+        "event.original": "device=\"SFW\" date=2017-01-31 time=14:52:11 timezone=\"IST\" device_name=\"CR750iNG-XP\" device_id=C44310050024-P29PUA log_id=138301618041 log_type=\"Sandbox\" log_component=\"Mail\" log_subtype=\"Allowed\" priority=Information user_name=\"\" src_ip= filename=\"\" filetype=\"\" filesize=0 sha1sum=\"\" source=\"\" reason=\"eligible\" destination=\"\" subject=\"\"",
+        "event.outcome": "success",
+        "event.severity": "6",
+        "event.timezone": "-02:00",
+        "event.type": [
+            "allowed",
+            "end",
+            "connection"
+        ],
+        "file.size": "0",
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/test/sandbox.log-expected.json:21]: should be a long not a string

File: x-pack/filebeat/module/sophosxg/firewall/test/sandbox.log-expected.json [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+        "event.module": "sophosxg",
+        "event.original": "device=\"SFW\" date=2017-01-31 time=14:52:11 timezone=\"IST\" device_name=\"CR750iNG-XP\" device_id=C44310050024-P29PUA log_id=138302218042 log_type=\"Sandbox\" log_component=\"Mail\" log_subtype=\"Denied\" priority=Critical user_name=\"jsmith@iview.com\" src_ip=10.198.47.112 filename=\"1.exe\" filetype=\"application/octet-stream\" filesize=153006 sha1sum=\"83cd339302bf5e8ed5240ca6383418089c337a81\" source=\"jsmith@iview.com\" reason=\"cached malicious\" destination=\"\" subject=\"\"",
+        "event.outcome": "success",
+        "event.severity": "2",
+        "event.timezone": "-02:00",
+        "event.type": [
+            "diened",
+            "connection"
+        ],
+        "file.extension": "application/octet-stream",
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/test/sandbox.log-expected.json:64]: this should be mappend to "file.mime_type" not "file.extension".  An extension would be "exe" or "png"

File: x-pack/filebeat/module/sophosxg/firewall/test/waf.log-expected.json [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+        "event.severity": "6",
+        "event.timezone": "-02:00",
+        "event.type": [
+            "diened",
+            "connection"
+        ],
+        "fileset.name": "firewall",
+        "http.request.method": "post",
+        "http.response.status_code": "401",
```
Comment (leehinman) [inline x-pack/filebeat/module/sophosxg/firewall/test/waf.log-expected.json:35]: should be a long not a string

**Comment by Bernhard-Fluehmann** [comment]
@StefanSa Thank you very much for your work. It would be great have this module available soon. I am working on a similar project for one of our customers. The  difference is that it needs to support Sophos UTM, especially the SG5xx product family. Do you know if these platforms are compatible in terms of logs or if two different modules are possible? If they are different, would it be possible to merge them together into one module or would it require two separate modules?

**Comment by StefanSa** [comment]
> @StefanSa Thank you very much for your work. It would be great have this module available soon. I am working on a similar project for one of our customers. The difference is that it needs to support Sophos UTM, especially the SG5xx product family. Do you know if these platforms are compatible in terms of logs or if two different modules are possible? If they are different, would it be possible to merge them together into one module or would it require two separate modules?

You do not only have to thank me alone, without the great support from elastic in particular from @P1llus (of course also from all the others here) it was difficult to do this.
I can't really tell you that because, i don't know the UTM syslog message.
If you would send me a few examples of UTM syslog messages, i can take a look at them.

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

**Comment by Bernhard-Fluehmann** [comment]
@StefanSa It seems that the structure of the logs is similar but unfortunately the log header and field names are different. Please have a look at [our repo](https://github.com/RealStuff/beats) . Under x-pack you find the [sophosutm module](https://github.com/RealStuff/beats/tree/sophosutm/x-pack/filebeat/module/sophosutm) which contains the current state of our work.
Even if the platforms are different, it would still be nice if they could be part of the same module. What do you think about? E.g. use sophos as module name and have dataset xg and utm instead of firewall? or Name the datasets firewall-xg and firewall-utm. Please let me know your thoughts.

**Comment by P1llus** [comment]
@Bernhard-Fluehmann I think in the long run, most likely for a separate release, we would need to think about renaming the module to sophos, and have the filesets be called sophosxg and sophosutm.

I might be able to convert the sophosxg one to follow that convention

**Comment by Bernhard-Fluehmann** [comment]
@P1llus  This is great, thanks. It seems that the official [namings](https://www.sophos.com/en-us/products/choose-firewall.aspx) are  is xg-firewall and sg-utm, while utm is the legacy platform and sg firewalls can be migrated to xg os.
Putting them into the same module as different datasets with own input config (like seen in the cisco module) would be great. This would allow for additional datasets like sophos email etc. in the future.
One thought about field name homologation. The most important fields are converted to ECS already, which makes both log types partially compatible. At the moment, non ECS fields would be different for both datasets. Since I expect more ECS fields to be available in the future, we can leave them alone for the moment. Since UTM seems to be legacy I can think to rename them to xg names as well if required.

BTW, I have some troubles building beats on my macbook. Since I have updated go and xcode command line tools make release works. But now the testsuite fails with the following error:
make: *** No rule to make target '../x-pack/filebeat/build/package/filebeat-darwin-amd64.dmg/beat-pkg-root/Library/Application', needed by 'filebeat.test'.  Stop.
make[1]: *** [system-tests-environment] Error 2
Do you know where the problem may be?

**Comment by P1llus** [comment]
@Bernhard-Fluehmann Yeah the pointer here will be to have it similar to the other modules, where the module name is the vendor and the fileset is the product. That way we can keep the separation between XG, UTM and any other module later on.

I think when this PR is merged, you can pull the latest update from our master, add your new fileset if you want, to the sophos module (il make sure to rename the folders before merging), and create a PR adding the fileset.

In terms of building, the way its built has changed a bit maybe, depending on what you are trying to do here.

If you are in the ./beats/x-pack/filebeat directory, make sure to delete the build folder and that you have the correct go version.
Currently this is built with 1.13.X, though I think the merges happening now is also going to 1.14.X.

I would recommend checking out what can be wrong, by ensuring that you have the correct go version and grab a clone of our repo and try to run it there.

If you are merging with master from time to time there might also be that you happen to be on a broken commit, should try to rebase with the newest master as well.

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

File: go.sum [hunk 22]
```diff
@@ -99,10 +99,8 @@ github.com/aerospike/aerospike-client-go v1.27.1-0.20170612174108-0f3b54da6bdc/g
 github.com/akavel/rsrc v0.8.0 h1:zjWn7ukO9Kc5Q62DOJCcxGpXC18RawVtYAGdz2aLlfw=
```
Comment (andrewkroh) [inline go.sum:99]: It looks like this does not introduce any Go changes so I would not expect any go.sum/go.mod changes in this PR.

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

**Comment by P1llus** [comment]
jenkins run tests

**Comment by adriansr** [comment]
jenkins, test this

**Comment by P1llus** [comment]
jenkins, test this

**Comment by P1llus** [comment]
jenkins, test this

**Pull request merged**